File: razer-sdk/src/main/java/com/bonepl/razersdk/animation/SimpleFrame.java
package com.bonepl.razersdk.animation;

import com.bonepl.razersdk.color.Color;
import com.bonepl.razersdk.color.StaticColor;
import com.bonepl.razersdk.sdk.RzKey;

import java.util.*;

/**
 * Single animation frame with one {@link Frame}
 */
public class SimpleFrame implements IFrame {
    private final Map<RzKey, Color> keysToColors = new EnumMap<>(RzKey.class);
    private boolean hasFrame = true;

    /**
     * Create an empty frame
     */
    public SimpleFrame() {
    }

    /**
     * Create a frame with all keys in the same color
     *
     * @param color {@link Color} of the fill
     */
    public SimpleFrame(Color color) {
        Arrays.stream(RzKey.values()).forEach(rzKey -> keysToColors.put(rzKey, color));
    }

    /**
     * Create a frame with one key in provided color
     *
     * @param rzKey {@link RzKey} to color
     * @param color {@link Color} of the fill
     */
    public SimpleFrame(RzKey rzKey, Color color) {
        keysToColors.put(rzKey, color);
    }

    /**
     * Create a frame with all provided keys in the same color
     *
     * @param rzKeys collection of {@link RzKey} to color
     * @param color  {@link Color} of the fill
     */
    public SimpleFrame(Collection<RzKey> rzKeys, Color color) {
        StaticColor nextColor = color.getColor();
        rzKeys.forEach(rzKey -> keysToColors.put(rzKey, nextColor));
    }

    /**
     * Create a frame from {@link RzKey} to {@link Color} map
     *
     * @param keysToColors {@link Map} of keys and corresponding fill color
     */
    public SimpleFrame(Map<RzKey, ? extends Color> keysToColors) {
        this.keysToColors.putAll(keysToColors);
    }

    /**
     * Check if frame is available
     *
     * @return true if {@link #getFrame} hasn't been called yet,
     * false otherwise
     */
    @Override
    public boolean hasFrame() {
        return hasFrame;
    }

    /**
     * Return the frame
     *
     * @return the single {@link Frame}
     * @throws NoSuchElementException when there is no frames available
     */
    @Override
    public Frame getFrame() {
        if (hasFrame) {
            hasFrame = false;
            return new Frame(keysToColors);
        } else {
            throw new NoSuchElementException("Animation does not have any frames to return");
        }
    }
}

File: src/razer-sdk/animation/SimpleFrame.ts
import {Color} from '../color/Color.js';
import {StaticColor} from '../color/StaticColor.js';
import {getRzKeyValues, RzKey} from '../sdk/RzKey.js';
import {Frame} from './Frame.js';
import {IFrame} from './IFrame.js';

/**
 * Single animation frame with one Frame
 */
export class SimpleFrame implements IFrame {
    private readonly keysToColors: Map<RzKey, Color> = new Map();
    private _hasFrame = true;

    constructor();
    constructor(color: Color);
    constructor(rzKey: RzKey, color: Color);
    constructor(rzKeys: readonly RzKey[], color: Color);
    constructor(keysToColors: Map<RzKey, Color>);
    constructor(
        colorOrKeyOrKeysOrMap?: Color | RzKey | readonly RzKey[] | Map<RzKey, Color>,
        color?: Color
    ) {
        if (colorOrKeyOrKeysOrMap === undefined) {
            // empty frame
        } else if (colorOrKeyOrKeysOrMap instanceof Map) {
            for (const [key, val] of colorOrKeyOrKeysOrMap) {
                this.keysToColors.set(key, val);
            }
        } else if (Array.isArray(colorOrKeyOrKeysOrMap)) {
            const nextColor: StaticColor = color!.getColor();
            for (const rzKey of colorOrKeyOrKeysOrMap) {
                this.keysToColors.set(rzKey, nextColor);
            }
        } else if (typeof colorOrKeyOrKeysOrMap === 'number' && color !== undefined) {
            // RzKey (number enum) + Color
            this.keysToColors.set(colorOrKeyOrKeysOrMap as RzKey, color);
        } else {
            // Single Color — fill all keys
            const c = colorOrKeyOrKeysOrMap as Color;
            for (const rzKey of getRzKeyValues()) {
                this.keysToColors.set(rzKey, c);
            }
        }
    }

    hasFrame(): boolean {
        return this._hasFrame;
    }

    pollFrame(): Frame {
        if (this._hasFrame) {
            this._hasFrame = false;
            return new Frame(this.keysToColors);
        }
        throw new Error('Animation does not have any frames to return');
    }
}
